igvc_navigation/src/path_follower/smooth_control.cpp

```diff
@@ -126,81 +127,129 @@ Action SmoothControl::getAction(const RobotState& state, const RobotState& targe
   return Action{ target_velocity_, w, 0 };
 }
 
-RobotState SmoothControl::getTargetPosition(const nav_msgs::PathConstPtr& path, unsigned int path_index,
-                                            const RobotState& state) const
+RobotState SmoothControl::getTargetPosition(const nav_msgs::PathConstPtr& path, const RobotState& state,
+                                            const std::optional<RobotState>& target, size_t path_index) const
 {
-  // target position
-  double tar_x;
-  double tar_y;
-  double tar_yaw = 1.57;
+  // If already have a target, find the closest point along the current path to the current target
+  // If the closest point is too far,then break and find a new target
+  if (target)
+  {
+    // If we haven't reached the target
+    if (!reachedTarget(state, *target))
+    {
+      size_t target_index = findTargetOnPath(path, *target);
+      RobotState newTarget{ path->poses[target_index] };
+      // If the target is found, and the target isn't behind us, and the target is closer to target + 1 than we are
+      double state_dist_target_next = state.distTo(path->poses[target_index + 1].pose.position);
+      double target_dist_target_next = target->distTo(path->poses[target_index + 1].pose.position);
+      if (target_index != 0 && target_index >= path_index && target_dist_target_next < state_dist_target_next &&
+          distAlongPath(path, path_index, target_index) < lookahead_dist_)
+      {
+        return newTarget;
+      }
+    }
+  }
+  return acquireNewTarget(path, state, path_index);
+}
+
+void SmoothControl::getWheelVelocities(igvc_msgs::velocity_pair& vel_msg, const Action& action) const
+{
+  vel_msg.left_velocity = action.v - action.w * axle_length_ / 2;
+  vel_msg.right_velocity = action.v + action.w * axle_length_ / 2;
+}
+
+SmoothControl::SmoothControl(double k1, double k2, double axle_length, double simulation_frequency,
+                             double target_velocity, double lookahead_dist, double simulation_horizon,
+                             double target_reached_distance, double target_move_threshold)
+  : k1_{ k1 }
+  , k2_{ k2 }
+  , axle_length_{ axle_length }
+  , simulation_frequency_{ simulation_frequency }
+  , target_velocity_{ target_velocity }
+  , lookahead_dist_{ lookahead_dist }
+  , simulation_horizon_{ simulation_horizon }
+  , target_reached_distance_{ target_reached_distance }
+  , target_move_threshold_{ target_move_threshold }
+{
+  ros::NodeHandle pNh{ "~" };
+  target_pub_ = pNh.advertise<nav_msgs::Path>("smooth_control/target", 1);
+  closest_point_pub_ = pNh.advertise<nav_msgs::Path>("smooth_control/closest_point", 1);
+}
 
+RobotState SmoothControl::acquireNewTarget(const nav_msgs::PathConstPtr& path, const RobotState& state,
+                                           size_t state_index) const
```

Comment:
`state_index` goes from being called `path_index` to `path_index` again and then to `state_index`. Try to keep the same name for consistency.